movici_simulation_core/models/rules/expression.py

```diff
@@ -0,0 +1,347 @@
+
+    expr_type: ExpressionType
+    operator: ComparisonOperator
+    value: Union[float, bool]
+    attribute_name: Optional[str] = None
+
+    def evaluate(
+        self,
+        simtime: Optional[float] = None,
+        clocktime: Optional[float] = None,
+        attributes: Optional[dict[str, Any]] = None,
+    ) -> bool:
+        """Evaluate this comparison.
+
+        :param simtime: Simulation time in seconds since start
+        :param clocktime: Clock time in seconds since midnight
+        :param attributes: Dict mapping attribute names to values
+        :returns: True if comparison is satisfied
+        :rtype: bool
+        """
+        if self.expr_type == ExpressionType.SIMTIME:
+            left_value = simtime
+        elif self.expr_type == ExpressionType.CLOCKTIME:
+            left_value = clocktime
+        else:
+            attributes = attributes or {}
+            left_value = attributes.get(self.attribute_name)
+
+        return self.operator.evaluate(left_value, self.value)
+
+    def get_attribute_names(self) -> set[str]:
+        """Return set of attribute names referenced by this comparison.
+
+        :returns: Set of attribute names (empty if simtime/clocktime)
+        :rtype: set[str]
+        """
+        if self.expr_type == ExpressionType.ATTRIBUTE and self.attribute_name:
+            return {self.attribute_name}
+        return set()
+
+
+@dataclass
+class BooleanExpression:
+    """A compound boolean expression with AND/OR/NOT operators."""
+
+    operator: str
+    operands: list[Union["BooleanExpression", Comparison]] = field(default_factory=list)
+
+    def evaluate(
+        self,
+        simtime: Optional[float] = None,
+        clocktime: Optional[float] = None,
+        attributes: Optional[dict[str, Any]] = None,
+    ) -> bool:
+        """Evaluate this boolean expression.
+
+        :param simtime: Simulation time in seconds since start
+        :param clocktime: Clock time in seconds since midnight
+        :param attributes: Dict mapping attribute names to values
+        :returns: True if expression is satisfied
+        :rtype: bool
+        """
+        if self.operator == "NOT":
+            return not self.operands[0].evaluate(simtime, clocktime, attributes)
+        elif self.operator == "AND":
+            return all(op.evaluate(simtime, clocktime, attributes) for op in self.operands)
+        elif self.operator == "OR":
+            return any(op.evaluate(simtime, clocktime, attributes) for op in self.operands)
+        return False
+
+    def get_attribute_names(self) -> set[str]:
+        """Return set of all attribute names referenced in this expression.
+
+        :returns: Set of attribute names
+        :rtype: set[str]
+        """
+        names: set[str] = set()
+        for operand in self.operands:
+            names.update(operand.get_attribute_names())
+        return names
+
+
+ParsedCondition = Union[Comparison, BooleanExpression]
```

Comment:
Might be clearer to make this a base class/protocol instead, so that the interface methods can be defined properly